TPA air speed speed estimation instead of TPA delay (for wings)

limonspb opened the pull request:
`#ifdef USE_WING`
This is a replacement for:
https://github.com/betaflight/betaflight/pull/13662 and https://github.com/betaflight/betaflight/pull/13778
Thanks to @ledvinap for the extensive help and discussions [here](https://github.com/betaflight/betaflight/pull/13833)

# Brief description
PR adds calculations of wing/plane relative air speed depending on throttle and pitch angle using a simple physical model, or a bit more advanced one depending on CLI settings. This speed estimation is used to adjust TPA factor. The goal is to achieve a proper plane PIDs multiplier depending on the current relative air speed estimation (relative air speed along the nose-tail axis).

# The problem
For quads the TPA argument is simply throttle. For planes/wings it should be the relative air speed. (more speed - less pids; less speed - more pids). GPS 3D speed could be useful, but only for no wind weather. Air speed sensors might be a good alternative. However can't require users to install it.

Previous approach `speed ~= lpf(throttle  + gravity)` mimics the delay between throttle and wing speed. However the lpf behavior doesn't reflect accurately how speed changes with throttle. For example, when throttle raised from 0 to 1 vs when throttle dropped from 1 to 0. The LPF step response time is equal in both cases. But actual wing speed reaches zero (when throttle dropped to 0) much later than speed reaches 100% (when throttle raised to 100%). Using two separate LPF filters (for raising throttle vs decreasing throttle) is also not an option, because, for example, when dropping throttle from 100% to 50% the terminal speed will be achieved much faster than when dropping throttle to zero.

So LPF approach requires more heuristics to be introduced. Instead a simple physical model can be used.

## `set tpa_speed_est_type = BASIC` (default for wings)
A basic physical model for speed estimation will be used with three simple tuning parameters:
```
set tpa_speed_est_basic_delay = 1000 # in milliseconds
set tpa_speed_est_basic_gravity = 50 # in precents
set tpa_speed_est_max_voltage = 2520 # theoretical max voltage in V*100
```
`tpa_speed_est_basic_delay` is arbitrary delay between throttle and speed. Speaking precisely it's the time to reach half of terminal velocity when throttle is raised from 0 to 100% in horizontal flight.
`tpa_speed_est_basic_gravity` is a gravity effect. Terminal speed of diving down zero throttle in precents from terminal speed of horizontal flight full throttle (in precents).
`tpa_speed_est_max_voltage` is used to just scale the throttle with voltage drop, or different cell count use.

So it's similar to old tuning parameters, except these values are recalculated into a simple physical model, and actual acceleration is being calculated using this formula:
![image](https://github.com/user-attachments/assets/9e458bea-2f47-4a5e-b0a0-0dc6d592fb9b)
a - acceleration
t - throttle (0 -1)
TWR - thrust to weight ratio
g - gravitational acceleration (~9.81)
C_d - drag coefficient
m - mass
alpha - pitch angle.

But when `set tpa_speed_est_type = BASIC` user does not have to deal with these physical values, and tuning with only `tpa_speed_est_basic_delay` and `tpa_speed_est_basic_gravity` supposed to be easier.

`set tpa_speed_est_type = BASIC`
If plane oscillates at quick throttle ramp ups - decrease `tpa_speed_est_basic_delay`.
If plane oscillates at quick throttle cuts - increase `tpa_speed_est_basic_delay`.
If plane oscillates during dives down at zero throttle - increase `tpa_speed_est_basic_gravity`.
If plane is too loose during dives down with zero throttle - decrease `tpa_speed_est_basic_gravity`.

## `set tpa_speed_est_type = ADVANCED`
Advanced physical model for speed estimation will be used, that is more accurately reflects speed behaviour:
![image](https://github.com/user-attachments/assets/43b3fc9a-7c9f-4d62-9db9-d396ce7cfd05)
a - acceleration
t - throttle (0 -1)
V_pitch100 - prop pitch speed at 100% throttle
TWR - thrust to weight ratio
g - gravitational acceleration (~9.81)
C_d - drag coefficient
m - mass
alpha - pitch angle.

tuning parameters:
```
set tpa_speed_est_adv_mass = 1000 # mass in grams
set tpa_speed_est_adv_drag_k = 45 # drag coefficient
set tpa_speed_est_adv_twr = 200 # thrust to weight ratio in %
set tpa_speed_est_max_voltage = 2520 # max voltage in V*100
set tpa_speed_est_adv_prop_pitch = 370 # propeller pitch in inch * 100
set motor_kv = 1960 # motor KV - reusing existing parameter
```

ADVANCED is more accurate physical model, but it is harder to tune.
But if a user is willing to record blackbox with `set debug_mode = TPA`, `set gps_use_3d_speed=ON`, and with GPS in non-windy day, he can process this bbx file with a python script that will calculate all of these parameters: https://github.com/limonspb/wing-speed-estimation

User only needs to know the plane `mass`, and `max_voltage`. Max voltage is a theoretically maximum voltage he will be using on this wing.

I think for most of users a `BASIC` model will be enough. Advanced is a bit more accurate tho, here is an example from a real flight log:
![image](https://github.com/user-attachments/assets/1d162082-9344-4324-a82f-79a8fe821368)

# Some code explanations
Acceleration calculations in pid.c are quite straightforward.
Main math is happening in pid_init.c.
For `BASIC` model thrust to weight ratio is easy:
`twr = 1.0f / (gravityFactor * gravityFactor);`
But for mass to drag ratio i had to do some integration:
`massDragRatio = (2.0f/logf(3.0f)) * (2.0f/logf(3.0f)) * pidRuntime.tpaSpeedEst.twr * G_ACCELERATION * delaySec * delaySec;`
Then the model maximum theoretical speed needs to be calculated. This is needed, because TPA argument should be within the range 0 - 1.
For `BASIC` model its the terminal speed of diving straight down with 100% throttle.
So the acceleration equation is solved for V, where a = 0, alpha = 90, throttle = 1.0.

For `ADVANCED` model the maximum speed can be either terminal speed of dive down with zero throttle, or terminal speed of dive down with 100% throttle. Which one is bigger. For example, for heavy and weak models just falling down can be faster, than falling down with zero throttle, because spinning prop can slow down the craft.

The average motor output was used instead of `throttle` (for thrust calculations). This helps, because zero throttle RC command might not be actually "stop motors" depending on motor idle settings.
Also in differential thrust scenario the average motor throttle can be a lot different from rc throttle due to airmode. 

# Real life testing
Even with `BASIC` model i was able to completely eliminate the oscillations at hard throttle changes, and dives down. Can't really 'feel' the difference between `ADVANCED` and `BASIC`, but i assume some slower planes will need `ADVANCED`. With the LPF approach and high PIDs that wasn't really possible, even with [tpa_delay_decrease_ratio](https://github.com/betaflight/betaflight/pull/13833).

Remember, this is not a real speed estimation. All this is used to only calculate the TPA argument. So approximate parameter values (both for BASIC and ADVANCED) should also do the job just fine.

# Future improvements
1. `ADVANCED` model is for ESC active breaking = ON. Old school electrical planes fly without active breaking.
2. Both models take into account voltage sag. However gas planes don't have that. Leaving gas planes for the future.
3. Gas planes don't use ESCs/motors, so can't use `getAverageMotorOutput` to calculate current thrust.
4. Drag coefficient is changing with pitching/rolling. Current models does not take into account that. Maybe someone can help to introduce drag coefficient parameter change depending on roll/pitch/yaw rates. (@AKlenk13 ? :))
5. For discussion: python script for calculating `ADVANCED` parameters is not so user friendly. It be nice to have Configurator interface for that. However the calculations are quite heavy, require multiple cores, and still taking ~10 - 20 minutes in my PC. So probably with JavaScript that would not be possible. Online web service, or PID toolbox?
6. Barometer can be used to adjust physical model according to the current air pressure.

limonspb commented:
For discussion: python script for calculating ADVANCED parameters is not so user friendly. It be nice to have Configurator interface for that. However the calculations are quite heavy, require multiple cores, and still taking ~10 - 20 minutes in my PC. So probably with JavaScript that would not be possible. Online web service, or PID toolbox? @bw1129 ? :)

limonspb commented:
@haslinghuis , its not refactoring, it's more of a replacing the old feature with the new one :)

haslinghuis commented:
Yes stills sounds like refactoring - replacing old feature with updated feature :rofl: 

Which label you'd suggest?

limonspb commented:
> Yes stills sounds like refactoring - replacing old feature with updated feature 🤣
> 
> Which label you'd suggest?

refactoring is making the same functionality with a better code ;)
Id' say its just a feature

Quick-Flash commented:
Couldn't you also look at the accelerometer readings to help improve this estimate? Use some sort of filter where the accelerometer readings will be slowly filtered to match this estimate? This way you fix any drift from the accelerometer, but also get the benefit of the true acceleration as measured on the aircraft?

limonspb commented:
> Couldn't you also look at the accelerometer readings to help improve this estimate? Use some sort of filter where the accelerometer readings will be slowly filtered to match this estimate? This way you fix any drift from the accelerometer, but also get the benefit of the true acceleration as measured on the aircraft?

I tried to use acceleration directly, and there are two problems:
1. After integration it has a drift, but that's expected. Probably high pass filter can get rid of it.
2. The acceleration amplitude for some reason is much smaller than it actually is for fast throttle changes (when acceleration is big).  Feels like need to apply some sort of expo or multiplier x2 or x3 on top of accelerometer readings, when it gets high.

It probably could be done and sorted out. And use some sort of complimentary filter to combine accelerometer and this algorithm. Maybe future updates. It honestly pretty accurate as it is now, considering we just need it for TPA

AKlenk13 commented:
> > Couldn't you also look at the accelerometer readings to help improve this estimate? Use some sort of filter where the accelerometer readings will be slowly filtered to match this estimate? This way you fix any drift from the accelerometer, but also get the benefit of the true acceleration as measured on the aircraft?
> 
> I tried to use acceleration directly, and there are two problems:
> 
> 1. After integration it has a drift, but that's expected. Probably high pass filter can get rid of it.
> 2. The acceleration amplitude for some reason is much smaller than it actually is for fast throttle changes (when acceleration is big).  Feels like need to apply some sort of expo or multiplier x2 or x3 on top of accelerometer readings, when it gets high.
> 
> It probably could be done and sorted out. And use some sort of complimentary filter to combine accelerometer and this algorithm. Maybe future updates. It honestly pretty accurate as it is now, considering we just need it for TPA

Quite confused by the acceleration magnitude not being as expected. That makes no sense to me. Does it have a very low frequency low pass filter applied or something?

limonspb commented:
> > > Couldn't you also look at the accelerometer readings to help improve this estimate? Use some sort of filter where the accelerometer readings will be slowly filtered to match this estimate? This way you fix any drift from the accelerometer, but also get the benefit of the true acceleration as measured on the aircraft?
> > 
> > 
> > I tried to use acceleration directly, and there are two problems:
> > 
> > 1. After integration it has a drift, but that's expected. Probably high pass filter can get rid of it.
> > 2. The acceleration amplitude for some reason is much smaller than it actually is for fast throttle changes (when acceleration is big).  Feels like need to apply some sort of expo or multiplier x2 or x3 on top of accelerometer readings, when it gets high.
> > 
> > It probably could be done and sorted out. And use some sort of complimentary filter to combine accelerometer and this algorithm. Maybe future updates. It honestly pretty accurate as it is now, considering we just need it for TPA
> 
> Quite confused by the acceleration magnitude not being as expected. That makes no sense to me. Does it have a very low frequency low pass filter applied or something?

Might be that or I used wrong units or something. Didn't try to dig deep. But despite the amplitude was low, the drift after integration was very high. It will need a strong high pass filter. And then will need the algorithm from this PR as a baseline after a low pass. IDK if it can improve anything, this algorithm is not noisy.

ledvinap commented:
> 2\. The acceleration amplitude for some reason is much smaller than it actually is for fast throttle changes (when acceleration is big).  Feels like need to apply some sort of expo or multiplier x2 or x3 on top of accelerometer readings, when it gets high.

It is quite possible that IMU complementary filter is interfering (it slowly rotates orientation so that acceleration matches gravity)

AKlenk13 commented:
> > Quite confused by the acceleration magnitude not being as expected. That makes no sense to me. Does it have a very low frequency low pass filter applied or something?
> 
> Might be that or I used wrong units or something. Didn't try to dig deep. But despite the amplitude was low, the drift after integration was very high. It will need a strong high pass filter. And then will need the algorithm from this PR as a baseline after a low pass. IDK if it can improve anything, this algorithm is not noisy.
> 
> Will need it anyways with or without accelerometer.

Yeah I think trying to add accel stuff is premature and overcomplicating things at this point. Tbh it feels like we need data from various aircraft types to see if things break down in that case or if the model is flexible but still accurate enough.

limonspb commented:
> 
> Yeah I think trying to add accel stuff is premature and overcomplicating things at this point. Tbh it feels like we need data from various aircraft types to see if things break down in that case or if the model is flexible but still accurate enough.

As far as i know this one is already more flexible than other opensource FWs. Some just use throttle as TPA argument, the other is a step forward, but a regular LPF1 delay, without gravity.

Today I put GPS on my other plane (delta wing with pusher motor), got this (was a little windy day):
![image](https://github.com/user-attachments/assets/84929619-2dd1-4589-b20b-6882ad37bf2b)

Above is a simulation with best parameters found by `differential_evolution`.
I wasn't flying with these best parameters, instead i was flying with defaults `BASIC`, but felt still pretty good, because of the scailing to 0-1:
![image](https://github.com/user-attachments/assets/d3736934-d8d1-4ebb-ae71-ed639a2bb777)

limonspb commented:
PS: updated `getAverageMotorOutput` function to proper handle the motor stop setting, and added proper default value for `tpa_speed_est_max_voltage`

limonspb commented:
@ledvinap i made a few changes:
1. Instead of thrust to weight ratio parameter (for ADVANCED type) it's now just thrust in grams. TWR is calculated internally from mass and thrust. This allows to change the mass independently, for example user just want to fly with way heavier battery.
2. I noticed something funny is happening around stall speeds. Usually the estimation is smaller than the actual GPS speed. Not that it's very important, because around stall speeds the pids are just maxed out anyways. But i also noticed my wing almost doesn't slow down at pitch ~0 degrees and stall speed. Part of it is probably because the accel calibration is slightly off. But also it seems that at ~0 degrees pitch the wing slides down and forward maintaining 0 degrees pitch. Idk if that's a wing shape, something common or not. Anyways i've added a new parameter pitch offset, that is zero by default
3. `set tpa_speed_est_pitch_offset = 0 # degrees * 100`
With 15 degrees offset for this particular wing now the slow speeds are more realistic:
![image](https://github.com/user-attachments/assets/342d2ff6-ccb4-4d5d-a7bc-1693d789ef40)

python differential evolution can give up with these many parameters, it can find something funny like offset = 25+ degrees depending on the flight. So need to be careful using it.

Overall this PR flies good, bbx mode TPA works as expected both in ADVANCED and BASIC mode.
For my wings the prop pitch doesn't affect much. ADVANCED just almost converges to BASIC if we believe in differential_evolution in python. It will probably make a difference with low pitch wide props that can generate lots of thrust when stationary but loose it with speed.

So i'm quite confident in math in this PR. It's ready for code review.

Here are some logs (attached)
```
    DEBUG_SET(DEBUG_TPA, 1, lrintf(pitchAngleRad * 180.0 / M_PIf)); // (before offset)
    DEBUG_SET(DEBUG_TPA, 2, lrintf(t * 1000.0f));
    DEBUG_SET(DEBUG_TPA, 3, lrintf(pidRuntime.tpaSpeedEst.speed * 10.0f));
    DEBUG_SET(DEBUG_TPA, 4, lrintf(tpaArgument * 1000.0f));
```
[white_basic.BFL.txt](https://github.com/user-attachments/files/17093343/white_basic.BFL.txt)
[black_advanced.bbl.txt](https://github.com/user-attachments/files/17093344/black_advanced.bbl.txt)
[black_basic.bbl.txt](https://github.com/user-attachments/files/17093345/black_basic.bbl.txt)
[white_advanced.BFL.txt](https://github.com/user-attachments/files/17093347/white_advanced.BFL.txt)

![image](https://github.com/user-attachments/assets/994d8991-05b4-4c81-bbe6-cdc07f4b0a9d)

ledvinap commented:
> 2. I noticed something funny is happening around stall speeds. Usually the estimation is smaller than the actual GPS speed. Not that it's very important, because around stall speeds the pids are just maxed out anyways. But i also noticed my wing almost doesn't slow down at pitch ~0 degrees and stall speed. Part of it is probably because the accel calibration is slightly off. But also it seems that at ~0 degrees pitch the wing slides down and forward maintaining 0 degrees pitch. Idk if that's a wing shape, something common or not. Anyways i've added a new parameter pitch offset, that is zero by default.

I guess it is related to cambered wings - lift is generated at zero AoA (and zero pich). The equations assume that pitch is directly related to change in height (and resulting force), but that is not true for gliding plane (zero pitch, but slowly losing altitude). I'm afraid that angle offset is only partial fix - in reality it is speed dependent (effective angle of attack is important).

At least part of this is representable as offset (and is based on physics), so the parameter makes sense. Personally, I'd go with 'we dont model this part and its influence is small enough', but the decision is on you.

>    `set tpa_speed_est_pitch_offset = 0 # degrees * 100`
>    With 15 degrees offset for this particular wing now the slow speeds are more realistic:

Degrees * 100 is overkill. BF does use decidegrees in some places, so it may be possible choice. And using degrees should be well within algorithm accuracy. (Ideally we should implement some float16 for config. It would fix a lot of problems and float16 is now quite normalized - last discussion was more than 10 years ago)

>    It is still not ideal match, but its just GPS speed, not actual air speed to compare against. Plus we are not taking into account roll/pitch/yaw rates for estimations, and probably some other factors.

Pitot tube would help a lot ;-)

> Overall this PR flies good, bbx mode TPA works as expected both in ADVANCED and BASIC mode. BASIC mode is easy to tune by "feel" as intended. For my wings the prop pitch doesn't affect much. ADVANCED just almost converges to BASIC if we believe in differential_evolution in python. It will probably make a difference with low pitch wide props that can generate lots of thrust when stationary but loose it with speed.

The results are great. I did not expect to get so close to measurement.

AKlenk13 commented:
> @ledvinap i made a few changes:
> 
> 1. Instead of thrust to weight ratio parameter (for ADVANCED type) it's now just thrust in grams. TWR is calculated internally from mass and thrust. This allows to change the mass independently, for example when user wants to fly with a much heavier battery on the same plane (will be able to use PID profiles to change mass, altho unless mass changes like +100%, it's not that important for TPA estimations)
> 2. I noticed something funny is happening around stall speeds. Usually the estimation is smaller than the actual GPS speed. Not that it's very important, because around stall speeds the pids are just maxed out anyways. But i also noticed my wing almost doesn't slow down at pitch ~0 degrees and stall speed. Part of it is probably because the accel calibration is slightly off. But also it seems that at ~0 degrees pitch the wing slides down and forward maintaining 0 degrees pitch. Idk if that's a wing shape, something common or not. Anyways i've added a new parameter pitch offset, that is zero by default.
>    `set tpa_speed_est_pitch_offset = 0 # degrees * 100`
>    With 15 degrees offset for this particular wing now the slow speeds are more realistic:
>    ![image](https://private-user-images.githubusercontent.com/2925027/369764377-342d2ff6-ccb4-4d5d-a7bc-1693d789ef40.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMDU0NzAsIm5iZiI6MTcyNzEwNTE3MCwicGF0aCI6Ii8yOTI1MDI3LzM2OTc2NDM3Ny0zNDJkMmZmNi1jY2I0LTRkNWQtYTdiYy0xNjkzZDc4OWVmNDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTUyNjEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDc2NWMyYjYzMTUyMjE1MjU1NDg0MjY2NzA1MTEyY2FlZDdmZDFlMWJkYjIxZmEyN2RjMmI1YWVmMDE3ZjBiZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MG5ZJA1pVE0BlF57mML-Ss5JboTBaDaCRn1YctXY95c)
>    It is still not ideal match, but its just GPS speed, not actual air speed to compare against. Plus we are not taking into account roll/pitch/yaw rates for estimations, and probably some other factors.
> 
> python differential evolution can give up with these many parameters, it can find something funny like offset = 25+ degrees depending on the flight. So need to be careful using it.
> 
> Overall this PR flies good, bbx mode TPA works as expected both in ADVANCED and BASIC mode. BASIC mode is easy to tune by "feel" as intended. For my wings the prop pitch doesn't affect much. ADVANCED just almost converges to BASIC if we believe in differential_evolution in python. It will probably make a difference with low pitch wide props that can generate lots of thrust when stationary but loose it with speed.
> 
> I'm quite confident in math in this PR. But not quite sure in some decisions i made in terms of code. So i think it's ready for code review.
> 
> Here are some logs (attached)
> 
> ```
>     DEBUG_SET(DEBUG_TPA, 0, lrintf(tpaFactor * 1000)); // final PID multiplier
>     DEBUG_SET(DEBUG_TPA, 1, lrintf(pitchAngleRad * 180.0 / M_PIf)); // (plane pitch in degrees before offset)
>     DEBUG_SET(DEBUG_TPA, 2, lrintf(t * 1000.0f)); // throttle adjsuted with battery sag
>     DEBUG_SET(DEBUG_TPA, 3, lrintf(pidRuntime.tpaSpeedEst.speed * 10.0f)); // speed estimation in m/sec * 10
>     DEBUG_SET(DEBUG_TPA, 4, lrintf(tpaArgument * 1000.0f)); // speed estimation scaled to 0-1 range * 1000
> ```
> 
> [white_basic.BFL.txt](https://github.com/user-attachments/files/17093343/white_basic.BFL.txt) [black_advanced.bbl.txt](https://github.com/user-attachments/files/17093344/black_advanced.bbl.txt) [black_basic.bbl.txt](https://github.com/user-attachments/files/17093345/black_basic.bbl.txt) [white_advanced.BFL.txt](https://github.com/user-attachments/files/17093347/white_advanced.BFL.txt) ![image](https://private-user-images.githubusercontent.com/2925027/369767857-7de492ad-a970-471d-a6f8-54b3bdbf7c16.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMDU0NzAsIm5iZiI6MTcyNzEwNTE3MCwicGF0aCI6Ii8yOTI1MDI3LzM2OTc2Nzg1Ny03ZGU0OTJhZC1hOTcwLTQ3MWQtYTZmOC01NGIzYmRiZjdjMTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTUyNjEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzE0MGNmNGI2YjA5OGE0NmM1YTU0YWY4YTZiN2JmOTZmYjY1ZDJlNjZiOWUxODdmZWEwMjRiMGJmNTczNmNlYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xwekHFzFbQX_pVApL8c8pCgfoUhVswIUMqfIWITZsfY)
> 
> On a side note, i think if someone wants to use ADVANCED and differential_evolution script, its better to lock prop pitch, mass, thrust, kv to known values (can be approximate), and let the script to just find drag coefficient. Works pretty good. Here i just locked prop to 4.1in, mass 0.9kg, thrust 2.5kg, kv 1960, max voltage 25.2: (leaving only drag coefficient unknown). Disregard GPS speed spikes at 50 seconds, lost GPS satellites count there for a few seconds. ![image](https://private-user-images.githubusercontent.com/2925027/369784760-da05540b-f2e2-4f5b-beec-4556c1abc8df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMDU0NzAsIm5iZiI6MTcyNzEwNTE3MCwicGF0aCI6Ii8yOTI1MDI3LzM2OTc4NDc2MC1kYTA1NTQwYi1mMmUyLTRmNWItYmVlYy00NTU2YzFhYmM4ZGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTUyNjEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTEwNDEzY2I0OGRhZWUxN2RhYWQ0YTA1MjRjNGJhY2NmNjJjZjdlYWJhZTAzZGY2MTJhMTExYTk0NWZhNTk1MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yck3t-ZZhpuYv3k1-QLxO1BpZ7N89sSxxpwqNXmXlto)

Yeah at low speed the angle of attack will be very high (aka the pitch and velocity vector are not very closely aligned) so I'm not surprised you have some weirdness in that condition. 

Overall I'm very impressed with the results here! I definitely want to try it out on my Flik when that arrives. Makes me want to build something basic in the meantime as well to test it out.

haslinghuis commented:
@limonspb there is a potential issue with TPA in master. See https://github.com/betaflight/betaflight/pull/13931

KarateBrot commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+
+void tpaSpeedEstAdvancedInit(const pidProfile_t *pidProfile)
+{
+    pidRuntime.tpaSpeedEst.twr = (float)pidProfile->tpa_speed_est_adv_thrust/(float)pidProfile->tpa_speed_est_adv_mass;
```
Thrust to weight ratio misses the gravitational acceleration. (weight = mass * gravity).
```suggestion
    const float weight = pidProfile->tpa_speed_est_adv_mass * G_ACCELERATION;
    pidRuntime.tpaSpeedEst.twr = pidProfile->tpa_speed_est_adv_thrust / weight;
```

limonspb replied:
> Thrust to weight ratio misses the gravitational acceleration. (weight = mass * gravity).

in CLI the `tpa_speed_est_adv_thrust` and `tpa_speed_est_adv_mass` both just in grams. So if a plane is 1000g weight but 2000g max thrust, then twr is 2.0.

KarateBrot replied:
Measuring prop thrust as masses would be odd. I think you are referring to kilogramm-force instead. That is still a force in Newton but normalized to the weight of 1kg of mass. So 1kgf is equal to 9.8N. In that case I'm satisfied.

limonspb replied:
> Measuring prop thrust as masses would be odd. I think you are referring to kilogramm-force instead. That is still a force in Newton but normalized to the weight of 1kg of mass. So 1kgf is equal to 9.8N. In that case I'm satisfied.

it's just common when bragging about how powerful a motor is to say "it provides 2kg thrust" :) Nobody says "19.6 newtons" in the hobby, and all the thrust-amp consumption tables for motors are in grams.

In the pidloop to calculate thrust **acceleration** i'm doing doing:
`const float thrust = (throttle * throttle - throttle * tpa->speed / tpa->propMaxSpeed) * tpa->twr * G_ACCELERATION;`

KarateBrot commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+    pidRuntime.tpaSpeedEst.massDragRatio = mass / dragK;
+    pidRuntime.tpaSpeedEst.propMaxSpeed = 0.0004233f * propPitch * motorConfig()->kv * pidRuntime.tpaSpeedEst.maxVoltage;
+
+    const float maxFallSpeed = sqrtf(mass * G_ACCELERATION / dragK);
```
```suggestion
    const float maxFallSpeed = sqrtf(weight / dragK);
```

KarateBrot reviewed (commented):
Some of the CLI settings I commented above are very general properties of the quad. Putting them inside TPA for wings instead of in a more generally accessible place seems "wasted" to me because they could be used for other features too. (for example mass, prop pitch, max voltage and thrust).

Below, I mostly made suggestions for renaming the function names, because they don't just return the value but actually have a full blown calculation inside. Using "get" in the name makes it seem like we can carelessly use it to (efficiently) get a value.

One bigger physics problem though: the thrust-to-weight-ratio in your code misses gravity. No idea if this means you need to change the algorithm or not.

limonspb commented on src/main/fc/parameter_names.h:
```diff
@@ -58,14 +58,20 @@
 #define PARAM_NAME_TPA_CURVE_PID_THR0 "tpa_curve_pid_thr0"
 #define PARAM_NAME_TPA_CURVE_PID_THR100 "tpa_curve_pid_thr100"
 #define PARAM_NAME_TPA_CURVE_EXPO "tpa_curve_expo"
+#define PARAM_NAME_TPA_SPEED_EST_TYPE "tpa_speed_est_type"
+#define PARAM_NAME_TPA_SPEED_BASIC_DELAY "tpa_speed_est_basic_delay"
+#define PARAM_NAME_TPA_SPEED_BASIC_GRAVITY "tpa_speed_est_basic_gravity"
+#define PARAM_NAME_TPA_SPEED_EST_ADV_PROP_PITCH "tpa_speed_est_adv_prop_pitch"
+#define PARAM_NAME_TPA_SPEED_EST_ADV_MASS "tpa_speed_est_adv_mass"
+#define PARAM_NAME_TPA_SPEED_EST_ADV_DRAG_K "tpa_speed_est_adv_drag_k"
+#define PARAM_NAME_TPA_SPEED_EST_ADV_THRUST "tpa_speed_est_adv_thrust"
+#define PARAM_NAME_TPA_SPEED_EST_MAX_VOLTAGE "tpa_speed_est_max_voltage"
+#define PARAM_NAME_TPA_SPEED_EST_PITCH_OFFSET "tpa_speed_est_pitch_offset"
```
@ledvinap need your opinion on parameter names, please :)

ledvinap replied:
`_est_` is a bit redundant, but `basic`/`advanced` is useful.

limonspb commented:
@KarateBrot  @ledvinap  @haslinghuis 
did renaming based on Jan's review, and rebased.

haslinghuis commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,76 @@ void pidResetIterm(void)
+{
+    const tpaSpeedParams_t *tpa = &pidRuntime.tpaSpeed;
+
+    const float thrust = (throttle * throttle - throttle * tpa->speed / tpa->propMaxSpeed) * tpa->twr * G_ACCELERATION;
```
```suggestion
    const float thrust = (throttle - tpa->speed / tpa->propMaxSpeed) * tpa->twr * G_ACCELERATION;
```

haslinghuis replied:
@KarateBrot please check the math here :)

limonspb replied:
t * t - t * A is not equal to t - A :)

haslinghuis replied:
Yeah you are correct. Only works when t == 1.0f :rofl:

limonspb replied:
Mark is only going full throttle, i like it :)

haslinghuis commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+#ifdef USE_WING
+void tpaSpeedBasicInit(const pidProfile_t *pidProfile)
+{
+    const float gravityFactor =  pidProfile->tpa_speed_basic_gravity / 100.0f;
```
```suggestion
    const float gravityFactor = pidProfile->tpa_speed_basic_gravity / 100.0f;
```

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,76 @@ void pidResetIterm(void)
+
+    const float thrust = throttle * throttle * tpa->twr * G_ACCELERATION;
+    const float drag = tpa->speed * tpa->speed / tpa->massDragRatio;
+    const float gravity = G_ACCELERATION * sin_approx(pitchAngle);
```
(technically, you can precompute 'forward' vector from pitchOffset and do the matrix multiplication [quite simple, with all zeroes]. It will avoid $sin(sin^{-1}(x))$ and also avoid $sin^{-1}(sin(x))$ ambiguity (https://www.wolframalpha.com/input?i=plot+sin%5E-1%28sin%28x%29%29+) - IMO the problem is that you can't tell if flying upright or inverted - sin(pitch) is the same, gravity force is fine, but pitchOffset is added or subtracted from angle)

limonspb replied:
i see the problem... How about something like that:
`const float gravity = G_ACCELERATION * sin_approx(pitch + cos(roll) * pitchOffset);`?

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,76 @@ void pidResetIterm(void)
+
+    pidRuntime.tpaSpeed.speed += a * pidRuntime.dT;
+    pidRuntime.tpaSpeed.speed = MAX(0.0f, pidRuntime.tpaSpeed.speed);
+    const float tpaArgument = constrainf(pidRuntime.tpaSpeed.speed/pidRuntime.tpaSpeed.maxSpeed, 0.0f, 1.0f);
```
```suggestion
    const float tpaArgument = constrainf(pidRuntime.tpaSpeed.speed / pidRuntime.tpaSpeed.maxSpeed, 0.0f, 1.0f);
```

ledvinap commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+    const float delaySec = pidProfile->tpa_speed_basic_delay / 1000.0f;
+
+    pidRuntime.tpaSpeed.twr = 1.0f / (gravityFactor * gravityFactor);
+    pidRuntime.tpaSpeed.massDragRatio = (2.0f/logf(3.0f)) * (2.0f/logf(3.0f)) * pidRuntime.tpaSpeed.twr * G_ACCELERATION * delaySec * delaySec;
```
(for consistency)
```suggestion
    pidRuntime.tpaSpeed.massDragRatio = (2.0f / logf(3.0f)) * (2.0f / logf(3.0f)) * pidRuntime.tpaSpeed.twr * G_ACCELERATION * delaySec * delaySec;
```

ledvinap commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+    const float dragK = pidProfile->tpa_speed_adv_drag_k / 10000.0f;
+    const float propPitch = pidProfile->tpa_speed_adv_prop_pitch / 100.0f;
+    pidRuntime.tpaSpeed.massDragRatio = mass / dragK;
+    pidRuntime.tpaSpeed.propMaxSpeed = 0.0004233f * propPitch * motorConfig()->kv * pidRuntime.tpaSpeed.maxVoltage;
```
```suggestion
    pidRuntime.tpaSpeed.propMaxSpeed = (2.54f / 100 / 60) * propPitch * motorConfig()->kv * pidRuntime.tpaSpeed.maxVoltage;
```

ledvinap commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,61 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+
+    const float maxFallSpeed = sqrtf(mass * G_ACCELERATION / dragK);
+
+    const float a = dragK;
```
(maybe add comment that propeller model is quite limited far above pitch speed)

ledvinap commented on src/main/flight/pid.h:
```diff
@@ -329,6 +342,16 @@ typedef struct pidCoefficient_s {
 } pidCoefficient_t;
 
+typedef struct tpaSpeedParams_s {
+    float maxSpeed;
```
(This value is not used ? )

ledvinap reviewed (commented):
sorry, forget to finish review.

Code looks great, I added some minor comments

limonspb commented on src/main/flight/mixer.h:
```diff
@@ -147,3 +147,7 @@ bool isFixedWing(void);
 float getMotorOutputHigh(void);
+
+#ifdef USE_WING
+float getAverageMotorOutput(void);
```
this is a very good find actually. I need to think about it :) Wouldn't matter for single motor, but it is important for dual motors.

limonspb commented on src/main/flight/mixer.c:
```diff
@@ -426,6 +435,20 @@ static void applyMixToMotors(const float motorMix[MAX_SUPPORTED_MOTORS], motorMi
+                averageMotor += (motorConvertToExternal(motor[i]) - PWM_RANGE_MIN) / (float)PWM_RANGE;
+            }
+        }
+        averageMotor /= mixerRuntime.motorCount;
```
for advanced model we need to know both throttle^2 and throttle. And in case of multiple motors to be accurate we need to know throttle^2 and throttle for each of the motors individually.

So i guess i will have to get rid of just average motor output function, and do the math on motors in PIDloop?

ledvinap replied:
You can handle BASIC by summing thrust^2 and than taking sqrt of result.

For dual propeller, you can compute total force (prom propeller model; using actual speed (pitch speed ratio). But in this situation, it quite possible that one propeller will have pitch speed considerably lower that plane speed. And currently used model does not work very well in this mode.

IMO we can ignore this problem in this PR. And in subsequent improvements, I'd try to improve model used (there is a lot of wind turbine theory available. so it should be relatively easy) and then fix/improve this.

limonspb replied:
@ledvinap i am a little lost here, could you plz help. Because we also want to adjust motor outputs with the voltage sag by a factor of `getBatteryVoltageLatest() / 100.0f / pidRuntime.tpaSpeed.maxVoltage`.

limonspb commented:
@ledvinap  @KarateBrot  @haslinghuis 
rebased and addressed some concerns.

### Single function for wing acceleration calculation
As Petr suggested, now using `inversePropMaxSpeed` which is 0 for `BASIC`, and 1/propMaxSpeed for ADVANCED.
It allowed to merge acceleration calculations into a single function `calcWingAcceleration()` instead of having separate functions for `BASIC` and `ADVANCED`. It also allows to use `ADVANCED` model with infinite prop pitch (set tpa_speed_adv_prop_pitch = 0).

### pitch offset changes
Pitch offset for gravity calculations is a bit controversial.
I decided to multiply pitchOffset by `cos(pitch) * cos(roll)`.
That way when flying inverted it's now making pitch offset in the opposite direction from when flying normal.
Also when flying nose up/down the pitch offset will be ->0.

### root mean square of motors outputs
Instead of `getAverageMotorOutput()` it is now using `getRootMeanSquareMotorOutput()` which is
![image](https://github.com/user-attachments/assets/d890eba1-bb75-441b-94a5-55f135ed454f)
then it's adjusted with battery factor in pid.c
now It's accurate for `BASIC` model even for more than 1 motor.
for `ADVANCED` model it is less accurate when there are multiple motors. But should be fine for a single motor.

limonspb commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,76 @@ void pidResetIterm(void)
+{
+    float batteryThrottleFactor = 1.0f;
+    if (pidRuntime.tpaSpeedEst.maxVoltage > 0.0f) {
+        batteryThrottleFactor = getBatteryVoltageLatest() / 100.0f / pidRuntime.tpaSpeedEst.maxVoltage;
```
yeah the > 1.0 values will make things more complicated. Need to know the max speed in order to get the speed in the range 0...1.

The idea is that user can set max voltage to the possible max voltage he will be using. And then he can put a smaller voltage battery without any adjustments to the tune.

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,60 @@ void pidResetIterm(void)
-    tpaArgument = pt2FilterApply(&pidRuntime.tpaLpf, tpaArgument);
-    DEBUG_SET(DEBUG_TPA, 2, lrintf(tpaArgument * 1000.0f));
+    const float thrust = (throttle * throttle - throttle * tpa->speed * tpa->inversePropMaxSpeed) * tpa->twr * G_ACCELERATION;
+    const float drag = tpa->speed * tpa->speed / tpa->massDragRatio;
```
(`*  tpa->inverseMassDragRatio`)

limonspb replied:
i guess it would be then `dragMassRatio`.
why would it be better here instead of division?

KarateBrot replied:
Division takes up to 14 CPU cycles. Multiplication only takes 4

ledvinap replied:
It was more about allowing zero drag, similar to pitch speed. But there is slight improvement in performance too

KarateBrot commented on src/main/flight/mixer.c:
```diff
@@ -473,6 +482,20 @@ static void applyMixToMotors(const float motorMix[MAX_SUPPORTED_MOTORS], motorMi
+            }
+        }
+    }
+    rootMeanSquareMotorOutput = sqrt(motorSumSquares / mixerRuntime.motorCount);
```
```suggestion
    rootMeanSquareMotorOutput = sqrtf(motorSumSquares / mixerRuntime.motorCount);
```

ledvinap replied:
Isn't there gcc warning for double math? Or it is only for double constants?

limonspb replied:
yeah im not getting any warnings if i use sqrt instead of sqrtf

haslinghuis replied:
Sorry to miss this one. @limonspb maybe check this in your announced follow up PR's

haslinghuis replied:
Also see: https://stackoverflow.com/questions/7431040/difference-between-sqrt-and-sqrtf

haslinghuis replied:
TL:DR using sqrt works with floats as sqrtf works also using double, long double and float

haslinghuis replied:
Anyway could update in follow PR

limonspb replied:
> Anyway could update in announced followup PR's if needed.

yeah i changed to `sqrtF` in this PR :)

ledvinap replied:
Most of SO answer is C++ only. In C, argument must be converted to `double`, then `sqrt` is calculated, using software implementation of `double`. This will be considerably more expensive than 14cycle VSQRT inctruction.

KarateBrot commented on src/main/flight/mixer.c:
```diff
@@ -438,6 +438,15 @@ static void applyRpmLimiter(mixerRuntime_t *mixer)
 #endif // USE_RPM_LIMIT
 
+#ifdef USE_WING
+static float rootMeanSquareMotorOutput = 0.0f;
+
+float getRootMeanSquareMotorOutput(void)
+{
+    return rootMeanSquareMotorOutput;
```
```suggestion
static float motorOutputRms = 0.0f;

float getMotorOutputRms(void)
{
    return motorOutputRms;
```

limonspb replied:
thanks, a good suggestion, done

KarateBrot reviewed (commented):
Just a suggestion for a shorter name

limonspb commented:
> Just a suggestion for a shorter name

thanks! Done!

ctzsnooze reviewed (approved):
Neat tidy coding.  Haven't flown but looks good to me.

haslinghuis commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,60 @@ void pidResetIterm(void)
 
     return tpaArgument;
 }
 #endif // #ifndef USE_WING
```
```suggestion
#endif // #ifdef USE_WING
```

haslinghuis replied:
@limonspb

limonspb replied:
fixed

haslinghuis commented on src/main/flight/pid.c:
```diff
@@ -287,23 +293,60 @@ void pidResetIterm(void)
+    return getMotorOutputRms() * batteryThrottleFactor;
+}
+
+
```
extra line
```suggestion
```

haslinghuis commented on src/main/flight/pid.c:
```diff
@@ -321,14 +364,15 @@ float getTpaFactorClassic(float tpaArgument)
     return 1.0f - tpaRate;
 }
 
+
```
extra line
```suggestion
```

haslinghuis commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,70 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
 #endif
 }
 
+
```
extra line
```suggestion
```

haslinghuis commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,70 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+    pidRuntime.tpaSpeed.maxSpeed = sqrtf(massDragRatio * pidRuntime.tpaSpeed.twr * G_ACCELERATION + G_ACCELERATION);
+    pidRuntime.tpaSpeed.inversePropMaxSpeed = 0.0f;
+}
+
```
extra line
```suggestion
```

haslinghuis commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,70 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+    pidRuntime.tpaSpeed.maxSpeed = MAX(maxFallSpeed, maxDiveSpeed);
+    UNUSED(pidProfile);
+}
+
```
extra line
```suggestion
```

haslinghuis commented on src/main/flight/pid_init.c:
```diff
@@ -66,6 +68,70 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
+}
+#endif // USE_WING
+
+
```
extra line
```suggestion
```

limonspb replied:
i thought we encourage two lines between function definitions?

haslinghuis replied:
Not I'm aware off. And if so it is not being used consistently :stuck_out_tongue: 

Can you point out where this was encouraged?

limonspb replied:
i guess it was only in my imagination. fixed :)

demvlad commented:
Yes, the airplanes perfomance depends from flight condition.
Very interest look how this work in really.
Because Very much hardly-defined data are used in the formulas.
The science can not compute props motor forces very exactly till.
The Cd is not const in really, it depends from angle of attack by parabola dependency. On high speed in horizont flight the angle of attack is small, the inductive drag is small too and Cd is const approximaly. But on midle-low speed in HF or on curves traectoryes the angle of attack is big and inductive drag increases very hardly. The drag is very big on stall mode, when angle attack is most high.
But we do not know angle of attack and real aerodynamic performance.
The angle of inclination of the trajectory (true pitch) what needs for accel is depend from variable real angle of attack too, but the const pitchOffset value is used instead of.

But these are very interest! Thank's!
Will test.

demvlad commented:
The term tuning by speed really need for airplanes. But to be honest, i suggest to do select switch for users, what use: estimators speed or gps speed. I understand, that gps speed is not air speed, but...

limonspb commented:
@ledvinap @demvlad  happy to report that airspeed estimation works nicely also on a classic RC airplane:
![image](https://github.com/user-attachments/assets/7a006610-db13-4753-9966-b5bd492ec90a)
![image](https://github.com/user-attachments/assets/5ba51621-8091-4254-9703-463d7651f19f)

ledvinap commented:
@limonspb : I'm always surprised how well it works ..

demvlad commented:
Considering what was written earlier, this is fine result. But it needs much testing on different flight modes.

demvlad commented:
@limonspb 
About angle of attack drag force influence.
In linear area the static angle of attack value has linear dependency from elevators deflection. The inductive drag from angle of attack is bigger (close to square dependancy), then self elevators drag. You can tray to use drag force addition what is proportionale square elevators deflection (or pids pitch output) to drag polar imitations.
It will help, maybe :)
But we never get just accuracy solution :)

demvlad commented:
![TPA](https://github.com/user-attachments/assets/bb0b20af-d8a4-416c-8ec1-39abf360911c)

limonspb commented:
> @limonspb The log looks very interest. If gps speed is true (Sat num 14...16!), and no wind. The V error is <0 in general - Vestim<Vgps most time. But what is increadeble: there are no big error accumulation, it has compensation every time. That mean - estimators formulas have different errors - sometime it has positive sign, sometime - negative. I try to describe behavior in max error point. t=01:44 At start t=01:30, close to horizon flight, the all are good - a small V error, Vest Vgps curves are paralel, estimation is good. At 01:33 you up throttle and start climb. The climb pitch is 60, the z accel is close to 1g. But we can look big estimators error on this climb! Cimputed dV/dt is bad, The V error grows up to -13m/s in climb end. This is 65% error!, relative gps speed. This is very big. What maybe reasone? Hardly to say, maybe motors force (after throttle push), maybe drag force... You stoped climb at 01:43 and pushed pitch stick down to get dive. During these short time 01:43-01:46 while you was moving stick, the speed error was decreasing very well! This mean the estimations error in other side! The airplanes speed accel was much less then estimator. What may be reasone? The accel z had big negative value -4g these short time, and big negative angle of attack was too. The airplane had big inductive drag probably, but yours estimator does not have it. It worked wrong, but previous big speed error was improved and there are very funny! The next time 1:46-1:50 you had 60grad pitch dive, the speed error was const - the estimators worked normal. Finelly, during climb and dive, the estimator had 2 big opposite sign error. And normal speed output in begin and end.

Are you looking at the log i shared in discord? Or is that your own flight?
If you looking at the log i shared, that log i used to find the best parameters for estimator, so it doesn't have them installed yet.

If it's your log, plz share in discord for example, ill check.

demvlad commented:
@limonspb 
this is yours log from discord.
I understood, this is some yours start point.
But it was interest for me.
Will interest look yours next logs in future.

I do not fly in winter :(. 
The build my foam airplane is in process (instead of crashed), will ready to spring, i think:)

demvlad commented:
@limonspb 
Sorry, one question, because i confused a small bit.
If i remember true, estimator computes (approximaly ofcourse) airplane airspeed (the speed relative real air around airplane, the X body axis component), and will work also in wind condition. But for example. I start airplane (armed, ready to airplane drop) with headwind 5m/s. What estimators air speed will i get at start (gps speed is zero) and in next time (forward horizon flight, Vgps=20m/s for example)?
Sorry if i missed something in code, but i see start estimators speed value as zero, and no wind velocity input.

demvlad commented:
The all are very interest for me anyway.
I respect yours big job.
But please, could you explain. Sorry if i missed something or do not understand.

Let's assume the all forces formulas are perfect, hypothetically..
1. No wind condition. I drop my plane hard with some initial speed.
    How estimator allow this speed?
2. I start plane with some head wind. The initial air speed is equal winds speed.
    The estimators initial speed is zero in program code.
    When estimator get me true (aproximally ofcourse) air speed value?

I understand good, that is not easy matter, to create the air speed estimator.
But i want to know its behavior on my airplane by using.

limonspb commented:
> The all are very interest for me anyway. I respect yours big job. But please, could you explain. Sorry if i missed something or do not understand.
> 
> Let's assume the all forces formulas are perfect, hypothetically..
> 
> 1. No wind condition. I drop my plane hard with some initial speed.
>    How estimator allow this speed?
> 2. I start plane with some head wind. The initial air speed is equal winds speed.
>    The estimators initial speed is zero in program code.
>    When estimator get me true (aproximally ofcourse) air speed value?
> 3. 1 and 2 together
> 
> I understand good, that is not easy matter, to create the air speed estimator. But i want to know its behavior on my airplane by using.
> 
> With best regards.

Case 1: initial speed will be zero till you give it some throttle, or tilt nose down.
Case 2: Initial speed will be zero, again, unless you give some throttle.
Case 3: same as 1 and 2.

Current estimator takes into account only throttle and pitch angle. So, for example, if you give full throttle while holding plane in your hands, the speed estimation will grow up despite you still holding plane in your hands.
Or if you hold you plane in your hands with zero throttle but tilt it down, the speed estimator will gradually increase "speed".
Or if you launching the plane from the fast moving train with very high initial speed, there is no way for the current estimator to know this initial true speed.

demvlad commented:
@limonspb , Thank you.
Yes, in addition, we have starts speed errors in case of up-down airplane nose during start process.
Seems i know how to improve any starts drop errors and get real wind speed also. :) lol. We should make forward horizon flight after start to achieve maximal speed for some fixed throttle. And hold maximal speed some time. If estimators formulas are good, theoreticaly, the airplane and estimator will have identity maximal speed (after different time ofcourse, we should wait this). From this point the estimator get true airspeed, and we can compute wing speed as difference estimators and gps speed. :)

As we can see, the estimator have errors much different types. But it has not any true signal correction. The corrections factors are opposite sign errors only. ....and horizon flight with some const speed during some time too, in case of fine forces formulas.
 
But i can not require estimators aqqurancy, because do not know how to get it. Yes, we can not use body accel, because can not transform body accel to air speed reference system. We need angle of attack and side angle to get this transform matrix, but we have not these sensors. 

I am not in opposite for air speed estimation experiments. But, is estimator has enough air speed accuracy to use it  instead of ground gps speed always? Will interest to look at yours best tuning flight logs at different modes what i wrote early.

Look at yours all airplane improvements i agree with you. The airplanes are much different device then multirotors drone. The airplanes properties have much dependencies from flight conditions. It has different control property and different maximal rates. The pids and required maximal rates tuning during flight is realy need for PID based airplane flight control system. 
Yes, some airplane property depend from air speed really, is not from ground speed. 

But i doubt that using air speed estimator give much accuracy today.
As i wrote early,  I suggest add option of using 3D gps speed for yours airplane pid controllers improvements. To have choice of speed source: experimental air speed with unknown accuracy or known accuracy gps 3D ground speed.

The real airplane have very many sensors, but its use pito pipe as air speed source...

limonspb commented:
> @limonspb , Thank you. Yes, in addition, we have starts speed errors in case of up-down airplane nose during start process. Seems i know how to resolve any starts drop errors and get real wind speed also. :) lol. We should make forward horizon flight after start to achieve maximal speed for some fixed throttle. And hold maximal speed some time. If estimators formulas are good, theoreticaly, the airplane and estimator will have identity maximal speed (after different time ofcourse, we should wait this). From this point the estimator get true airspeed, and we can compute wing speed as difference estimators and gps speed. :)
> 
> As we can see, the estimator have errors much different types. But it has not any true signal correction. The corrections factors are opposite sign errors only. ....and horizon flight with some const speed during some time too, in case of fine forces formulas.
> 
> But i can not require estimators accuracy, because do not know how to get it. Yes, we can not use body accel, because can not transform body accel to air speed reference system. We need angle of attack and side angle to get this transform matrix, but we have not these sensors.
> 
> I am not in opposite for air speed estimation experiments. But, is estimator has enough air speed accuracy to use it instead of ground gps speed always? Will interest to look at yours best tuning flight logs at different modes what i wrote early.
> 
> Look at yours all airplane improvements i agree with you. The airplanes are much different device then multirotors drone. The airplanes properties have much dependencies from flight conditions. It has different control property and different maximal rates. The pids and required maximal rates tuning is realy need for PID based airplane flight control system. Yes, some airplane property depend from air speed really, is not from ground speed.
> 
> But i doubt that using air speed estimator give much accuracy today. As i wrote early, I suggest add option of using 3D gps speed for yours airplane pid controllers improvements. To have choice of speed source: experimental air speed with unknown accuracy or known accuracy gps 3D ground speed. And definitally, the gps speed will usefull in unwind flight conditions.
> 
> The real airplanes have very many sensors, but its use pito pipe as air speed source...

Oh i can tell you 100% this airspeed estimation is really a savior for tight PID tune. Remember, it's not about knowing the true airspeed value (altho it be nice), it's just to know the PID multiplier (calculated from airspeed estimation and then hyperbolic TPA curve applied). First i did a classic throttle-based PID attenuation like Ardupilot/Inav have. And i could not tune pids tightly: everything good in horizontal flight, but falls apart on dives down with zero throttle, for example. Or very unstable at nose up vertical hovers, because throttle is high and PIDs become low. Or on fast throttle changes i was having oscillations, but at slow throttle changes everything was good. All this is all solved with the current speed estimation, altho it's not ideal yet.

Here is a history of all my attempts, if you are interested:
https://github.com/betaflight/betaflight/pull/13662
https://github.com/betaflight/betaflight/pull/13778
https://github.com/betaflight/betaflight/pull/13833
and finally current PR:
https://github.com/betaflight/betaflight/pull/13895

Plus hyperbolic curve, of course:
https://github.com/betaflight/betaflight/pull/13805

demvlad commented:
Thank's for details. I studied yours PRs. The all are actually.
But this pr had most questions for me:)
Yes, of course, 20mph wind :), the starts drop errors and simmilar has no matter :)
And maximal flight velocities is good for estimators errors corrections... But in low wind or without it, the errors maybe greate then wind.
Thank's for attention. Will look next flights.
Good luck!

demvlad commented:
Sorry, some though about estimators testing.
I think  will good test first the forward horizon flight with several fixed throttle values in midle-maximal speed range..
This is simple, but need to hold horizon flight accuratly:
1. Max throttle, get maximal speed, hold it some time.
2. Move throttle back on some step, wait const speed
3. Do (2) several time while const speed is exists on current fixed throttle.
Because on some speed (and throttle pos) you can not get const speed.
The speed will decelerate the all time in horizon flight.
These speed point is flight modes border:
1nd mode, the bigger speed need much throttle and much throttle has much speed. Less speed needs less throttle and less thrittle gives less fixed soeed
2nd mode, the all are opposite - the less speed needs more throttle, and less throttle gives " infinite" speed decceleration.
Definitelly, the current estimator (without inductive polar) will not work properly close this border speed and on 2nd flight mode,
but this border speed can be helpfull for inductive drag checking during flight simulation.
By compare airplane and estimators fixed speed points we can look at REAL estimators accuracy in static modes.
These points can be helpfull for estimators drag and motors forces checking and tuning by using mathematics simulation.

When these modes will normal, can be test next...
- the climbs and dives
- curves in horizon flight
- split-S, power loops
...

haslinghuis commented:
Perhaps introduce wind estimation

limonspb commented:
> Sorry, some though about estimators testing. I think will good test first the forward horizon flight with several fixed throttle values in midle-maximal speed range.. This is simple, but need to hold horizon flight accuratly:
> 
> 1. Max throttle, get maximal speed, hold it some time.
> 2. Move throttle back on some step, wait const speed
> 3. Do (2) several time while const speed is exists on current fixed throttle.
>    Because on some speed (and throttle pos) you can not get const speed.
>    The speed will decelerate the all time in horizon flight.
>    These speed point is flight modes border:
>    1st mode, the bigger speed need much throttle and much throttle has much speed. Less speed needs less throttle and less thrittle gives less fixed soeed
>    2nd mode, the all are opposite - the less speed needs more throttle, and less throttle gives " infinite" speed decceleration.
>    Definitelly, the current estimator (without inductive polar) will not work properly close this border speed and on 2nd flight mode,
>    but this border speed can be helpfull for inductive drag checking during flight simulation.
>    By compare airplane and estimators fixed speed points we can look at REAL estimators accuracy in static modes.
>    These points can be helpfull for estimators drag and motors forces checking and tuning by using mathematics simulation.
> 
> When these modes will normal, can be test next...
> 
> * the climbs and dives
> * curves in horizon flight
> * split-S, power loops
>   ...

I think you underestimate the work that has been done :) everything you talking about was done. I made countless flights and logs gradually leading to this PR, doing different maneuveres, different speeds. Easily 200+ flights before we end up in this code.

You just gotta actually fly and see yourself if it's working for you :) 

Ar this point I'm certain it will work as intended except in very hard maneuveres where you will get a bit less PIDs. But that in the works too.

demvlad commented:
I apologize again, but I wouldn't write if I didn't know how difficult it is to make an adequate reality model even for one aircraft...
Thank you for big job and improvements, its are really actualy for using PIDs control system on airplanes.
Good luck!
